### DISCUSSION TITLE
Running cocotb without Makefile

### DISCUSSION THREAD

**Discussion by PiyushSaini07** [cocotb/cocotb#3146] [Q&A] [answered]
Hi,

To run the tests I've been using Makefile by invoking `make TESTCASE=<testname>`.

Is there a way to bypass Makefile all together and directly use python to do so?
I tried out a few things from this [blog](https://ahmedalsawi.github.io/posts/2022/10/cocotb-internals-python-runner/), however it requires a lot more effort than directly using Makefile.

With more digging, I came across [cocotb-test](https://github.com/themperek/cocotb-test). cocotb-test seems to be what I'm looking for. I just wanted to know if there is any other method to run cocotb via python?

**Comment by ktbarrett** [comment, marked as answer]
See the [Extending Existing Build Flows](https://docs.cocotb.org/en/stable/custom_flows.html) section of the documentation. It's usually as simple as adding a flag to the simulator call to load the GPI and a couple environment variables (`MODULE` and `TOPLEVEL`) for the regression manager to find tests and an HDL toplevel.

**Reply by PiyushSaini07** [reply]
Thanks for the suggestion.
I tried the following steps:
1. set `LIBPYTHON_LOC` and `MODULE` environment variables.
2. ran the command `xrun -loadvpi $(cocotb-config --lib-name-path vpi xcelium):vlog_startup_routines_bootstrap -f <myFileList>`

xrun compiled the files successfully but it did not take the stimuli I wrote in python TB. Is there something am I missing?

Edit: Adding the log I got after running-

```
TOOL:	xrun	22.03-s004: Started on Nov 25, 2022 at 00:07:33 CST
xrun: 22.03-s004: (c) Copyright 1995-2022 Cadence Design Systems, Inc.
Loading snapshot worklib.tb_counter:sv .................... Done
SVSEED set from command line: 9999999
xmsim: *W,LIBLDFL: Failed to load dynamic library /data/projects/share_with_EDA/cadence_venv/lib64/python3.6/site-packages/cocotb/libs/libcocotbvpi_ius.so:
	/data/projects/share_with_EDA/cadence_venv/lib64/python3.6/site-packages/cocotb/libs/libcocotbvpi_ius.so: wrong ELF class: ELFCLASS64.
xcelium> source /data/CadenceInstall/xcelium2203/xcelium2203_s004/tools/xcelium/files/xmsimrc
xcelium> 
xcelium> database -open counter -into ./outputs/counter.shm -default
Created default SHM database counter
xcelium> probe -create -packed 262144 tb_counter -all -memories -depth all -database counter -waveform
Created probe 1
xcelium> run 10ms
Ran until 10 MS + 0
xcelium> exit
```

**Reply by ktbarrett** [reply]
> `wrong ELF class: ELFCLASS64`

Seems like you are trying to load 64-bit cocotb libraries into a 32-bit simulator. You will have to use a 32-bit installation of Python and install cocotb into that environment.

**Reply by cmarqu** [reply]
I guess the flag `-64` or alternatively the environment variable `CDS_AUTO_64BIT=all` is missing from the xrun call.

**Reply by PiyushSaini07** [reply]
Thanks for the help! Adding `-64` worked for me.
